rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+
+## Settings
+Provides HTTP API for Kibana UI to configure user-specific settings.
+Not provided in MVP, we might add in the future:
+- Unsubscribe from a source
+- Subscribe to a source
+
+#### Filtering
+Filtering allows users to hide notifications they aren’t interested in. Kibana UI attaches a list of filters when requesting notifications.
+Kibana stores filter as a part of User settings data to provide a seamless experience between different deployments on Cloud and user sessions.
+
+
+# Drawbacks
+
+- It is not entirely clear what functionality will be supported by UNS and when this will be known.
+It forces us to evaluate each feature separately as to how difficult it will be to adapt it to UNS later.
+- The service relies on the yet-to-be-added ability to uniquely identify the user, enumerate users with a given set of permissions.
+- The service doesn't provide a built-in security mechanism but relies on a notification source implementation.
+- Lack of a centralized i18n service makes message transferring over-bloated due to the necessity to send all the possible translations over the wire.
+
+
+# Alternatives
+
+### Threat Cloud as another pluggable storage mechanism.
+Kibana notification service collects internal notifications and shows them in the UI for both on-prem and Cloud environments.
+Kibana reuses Notification service implementation on Cloud but enhances it with additional pluggable storage of external notifications.
+![img](../images/0014/alternative_implementation.png)
+
+The main benefit of this approach is that Kibana always uses the same service in any environment.
+But in addition to the restrictions listed in the [Drawbacks](#drawbacks) section, the following are added:
+- Each retrieval of a notification list still requires waiting for a roundtrip to UNS.
+- The list of Kibana notifications stored in an instance might be out of sync with a copy stored in UNS due to network problems. 
+- Storing notifications in the Cloud deployment and their copy in UNS consumes twice the disk space.
+
+# Adoption strategy
+
+We start with introducing Kibana Notification Service as a Kibana plugin, adding more notification sources along the road.
+Integration with UNS is something that is not going to happen any time soon. The work can be divided into the following phases:
+
+### Phase I
+Kibana shows notifications created by the Newsfeed Kibana plugin in Notification flyout.
+![img](../images/0014/kns_flyout.png)
+A notification cannot have a particular recipient in the lack of a way to identify a user in the system. A notification will be shown to all the users instead.
+User-specific notification state (*read_status*) is not stored in Kibana but browser Local Storage.
+### Phase II
+When [user profiles](https://github.com/elastic/kibana/issues/17888) are supported, Kibana UI starts showing user-specific notifications.
```

Comment:
One concern I have is how we will handle the case of when security is disabled. How do you imagine the storage mechanism working in that case? Do we know if we could just not support notifications in that type of deployment?

Reply:
> Do we know if we could just not support notifications in that type of deployment?

This is my vote. The onboarding notifications potentially make sense w/o security, and I could make an argument for others, but this could also be an enticing forcing function to get our clusters secured.

We are still planning on enabling security by default starting in 8.0, so the burden of getting security setup will be greatly diminished.

Reply:
> One concern I have is how we will handle the case of when security is disabled. How do you imagine the storage mechanism working in that case? Do we know if we could just not support notifications in that type of deployment?

The storage mechanism doesn't depend on `user_id` as long as a notification source passes it during notification creation. 

We need `user_id` when retrieving notifications addressed to a particular recipient.
Lack of `user_id` is not an obstacle to show notifications addressed to all the users (referred to as `global` notifications in the RFC).

Also, the ability to disable Security plugin has been deprecated https://github.com/elastic/kibana/pull/85159

cc @alexfrancoeur as it is a product decision as well.

Reply:
> Also, the ability to disable Security plugin has been deprecated #85159

Note that enabling Kibana's security plugin `!==` having stack security features enabled. Future versions of the stack will still be able to run without security features -- they just won't be able to explicitly disable Kibana's security plugin anymore.

Reply:
I think we should prefer for "global" cluster notifications to still function even if the parts of our stack that allow us to differentiate users is unavailable. That being said, it isn't yet clear to me how complex bringing user identity into Kibana and this system will prove to be (very is my guess). So we should keep an eye on this and I'd like to understand the details of this better myself as we go.